### DISCUSSION TITLE
Deployment to production and risks for developers versus code of ethics ACM

### DISCUSSION THREAD

**Discussion by mick3yy** [community/community#58555] [Programming Help] [answered]
### Body

How to deal with untrustworthy live production project when your manager doesn't take your advice seriously.

## WHAT 
Recently I was asked to support live project that was mainly occupied by mid level developers, I'm my self a senior lead developer, before jumping in to the project I found major circumvention practices on client production code.

## WHERE
Some issues im happy to share:

No production environment, no git (source control) or pipeline deployment to production, all done manually from users machine to deploy code to production from staging environment.

This was one of major issues, since no one can track what is really send to production except for Project Managers trust in his developer.

The Project Manager, or the CEO does not want to adhere to my request to fix the issue, and they kinda try to force me to work on it.

If I don't do it, i may risk loosing my job, if I agree to do it I may risk my job and reputation there after if something goes wrong.

The company is large with Devop services and pipelines.

## HOW
Can someone provide any valuable information please, the company is in Thailand.

Here I found some information on:
[ACM ethics, Software Engineering Code.](https://ethics.acm.org/code-of-ethics/software-engineering-code/
)

For now I shared the information within the community of developers I know regarding  safety procedures and how important it is for experienced developers to check UAT to production is sound and secure.

I notified the CEO and PM about circumvented practices as well.

### Any suggestions?
What would you do in this situation, just follow the job requirements or acknowledge the problem and provide solution to resolve it before you gain trust working in new project ?

Thanks for your time
<br/>
Michael

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by airtower-luna** [comment, marked as answer]
Phew, that's a difficult situation! The first bit of advice is to do what's often called "cover your ass". As in, make sure they can't pin the blame on you if (or when, from what you described it's more a matter of when!) things blow up. When you warn them, do it in writing. When they give questionable instructions, make sure to get them in writing. Keep that paper trail, to prove your (and their) role if necessary. I can't say how much that helps if they want to fire you though, that'd depend on applicable labor law (a union representative or lawyer is probably the best address to ask about that).

What else you can do probably depends on applicable laws & regulations: Is what they want you to do "just" shoddy work, or illegal?

If it is "just" shoddy, I'm afraid other than protecting yourself, reporting to higher-ups in the company, and looking for a better job (both for your peace of mind and to be gone before things collapse) there's little you can do. You can warn them, but you can't force them to change their mind.

If it's *illegal*, things look different: You could give a hint to the appropriate authorities. I don't know what those look like in Thailand, but here (Germany/EU) there's a variety depending on topic. E.g. if users' private data isn't protected properly the state privacy official can force the company and fine them, or if a bank isn't keeping proper records the banking oversight authority can take action (a while ago I read reports about a banking startup being prohibited from opening new accounts until they got their measures against money laundering up to regulations).

The catch is that the manager/CEO/… might still try to get revenge on you, so for your own safety it's important to check on whistleblower protection rules in your jurisdiction: Who can you legally report to? Do you have to report internally first? Are you protected from mistreatment/firing for making the report? And even if you make sure everything you do is legal, and they can't fire you either, it might be questionable if you still want to work for the same people after…

Best wishes!

**Reply by mick3yy** [reply]
Hey thanks for such comprehensive response! I needed someone to see the issue the same way I do.

I will need to be more vigilant from now on
